// react-jwt-auth/lib/es6/components/Google.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _OAuth = require('./OAuth2');

var _OAuth2 = _interopRequireDefault(_OAuth);

var _enverse = require('enverse');

var _enverse2 = _interopRequireDefault(_enverse);

var _utils = require('../internals/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var defaultProps = {
	clientId: null,
	name: 'google',
	label: 'Sign in with Google',
	tokenEndpoint: '/auth/google',
	oauthProvider: 'google',
	oauthEndpoint: 'https://accounts.google.com/o/oauth2/auth',
	redirectUri: _enverse2.default.is.browser ? window.location.origin : undefined,
	requiredUrlParams: ['scope'],
	optionalUrlParams: ['display', 'state'],
	scope: ['profile', 'email'],
	scopePrefix: 'openid',
	scopeDelimiter: ' ',
	display: 'popup',
	oauthType: '2.0',
	popupOptions: { width: 452, height: 533 },
	style: {
		color: '#fff',
		backgroundColor: '#dd4b39',
		border: '1px solid #d54331',
		padding: '5px 15px'
	},
	state: function state() {
		return encodeURIComponent(Math.random().toString(36).substr(2));
	}
};

var Google = function (_Component) {
	_inherits(Google, _Component);

	function Google() {
		_classCallCheck(this, Google);

		return _possibleConstructorReturn(this, (Google.__proto__ || Object.getPrototypeOf(Google)).apply(this, arguments));
	}

	_createClass(Google, [{
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				_OAuth2.default,
				this.props,
				this.props.children || null
			);
		}
	}]);

	return Google;
}(_react.Component);

exports.default = Google;


Google.defaultProps = defaultProps;
Google.propTypes = _OAuth2.default.propTypes;
