// react-jwt-auth/lib/cjs/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.Auth = exports.Google = exports.Facebook = undefined;

var _shared = require('./shared');

var _local = require('./local');

var _user = require('./user');

var _Facebook = require('./components/Facebook');

var _Facebook2 = _interopRequireDefault(_Facebook);

var _Google = require('./components/Google');

var _Google2 = _interopRequireDefault(_Google);

var _Auth = require('./components/Auth');

var _Auth2 = _interopRequireDefault(_Auth);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Facebook = _Facebook2.default;
exports.Google = _Google2.default;
exports.Auth = _Auth2.default;
exports.default = {
	Auth: _Auth2.default,
	Facebook: _Facebook2.default,
	Google: _Google2.default,
	isAuthenticated: _shared.isAuthenticated,
	login: _local.login,
	logout: _local.logout,
	signup: _local.signup,
	getToken: _shared.getToken,
	setToken: _shared.setToken,
	refreshToken: _local.refreshToken,
	getAuthHeader: _shared.getAuthHeader,
	getProfile: _user.getProfile,
	updateProfile: _user.updateProfile
};
